docs/topics/addons.rst

```diff
@@ -0,0 +1,348 @@
+            try:
+                import boto
+            except ImportError:
+                raise RuntimeError("myaddon requires the boto library")
+            self.export_config(config, settings)
+
+Enable a component that lives relative to the add-on (see
+:ref:`topics-api-settings`)::
+
+    from scrapy.addons import Addon
+
+    class MyAddon(Addon):
+        name = 'myaddon'
+        version = '1.0'
+        component = __name__ + '.downloadermw.coolmw'
+        component_type = 'DOWNLOADER_MIDDLEWARES'
+        component_order = 900
+
+Instantiate components ad hoc::
+
+    from path.to.my.pipelines import MySQLPipeline
+
+    class MyAddon(object):
+        name = 'myaddon'
+        version = '1.0'
+
+        def update_settings(self, config, settings):
+            mysqlpl = MySQLPipeline(password=config['password'])
+            settings.set(
+                'ITEM_PIPELINES',
+                {mysqlpl: 200},
+                priority='addon',
+            )
+
+Provide add-on interface along component interface::
+
+    class MyPipeline(object):
+        name = 'mypipeline'
+        version = '1.0'
+
+        def process_item(self, item, spider):
+            # Do some processing here
+            return item
+
+        def update_settings(self, config, settings):
+            settings.set(
+                'ITEM_PIPELINES',
+                {self: 200},
+                priority='addon',
+            )
+
+Enable another addon (see :ref:`topics-api-addonmanager`)::
+
+    class MyAddon(object):
+        name = 'myaddon'
+        version = '1.0'
+
+        def update_addons(self, config, addons):
+            if 'httpcache' not in addons.enabled:
+                addons.add('httpcache', {'expiration_secs': 60})
+
+Check configuration of fully initialized crawler (see
+:ref:`topics-api-crawler`)::
+
+    class MyAddon(object):
+        name = 'myaddon'
+        version = '1.0'
+
+        def update_settings(self, config, settings):
+            settings.set('DNSCACHE_ENABLED', False, priority='addon')
+
+        def check_configuration(self, config, crawler):
+            if crawler.settings.getbool('DNSCACHE_ENABLED'):
+                # The spider, some other add-on, or the user messed with the
+                # DNS cache setting
+                raise ValueError("myaddon is incompatible with DNS cache")
+
+Provide add-on interface through a module:
+
+.. code-block:: python
+
+    name = "AddonModule"
+    version = "1.0"
+
+
+    class MyPipeline(object):
+        ...
+
+
+    class MyDownloaderMiddleware(object):
+        ...
+
+
+    def update_settings(config, settings):
+        settings.set(
+            "ITEM_PIPELINES",
+            {MyPipeline(): 200},
+            priority="addon",
+        )
+        settings.set(
+            "DOWNLOADER_MIDDLEWARES",
+            {MyDownloaderMiddleware(): 800},
+            priority="addon",
+        )
+
+Forward to other add-ons depending on Python version::
```

Comment:
This can be done in other ways; also, the example uses six.PY3, and we don't support Python 2 for a long time. I'd drop this feature and the example as well.